### PR TITLE
Python packages set: major updates

### PR DESCRIPTION
**Pull request by FRidh**
###### Motivation for this change

It's been a long time since we did major updates in the whole set. Applications were not automatically updated, but likely need some fixing as well.

https://hydra.nixos.org/jobset/nixpkgs/python-unstable

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by FRidh** [comment]
numpy needs https://github.com/numpy/numpy/pull/17776

**Comment by jonringer** [comment]
the use of flit_core is making gimp fail because python2Packages.flit_core is disabled, which failing the eval as gimp is a darwin constituent https://gist.github.com/GrahamcOfBorg/f0aa782e418d7d426a7bf7a5a7372186

**Comment by jonringer** [comment]
maybe we should revert the packaging bump until they figure out their existence.

File: pkgs/development/python-modules/nbconvert/default.nix [hunk 1]
```diff
@@ -29,21 +30,39 @@ buildPythonPackage rec {
...
   '';
 
+  pytestFlagsArray = [
+    "--ignore=nbconvert/preprocessors/tests/test_execute.py"
+    # can't resolve template paths within sandbox
+    "--ignore=nbconvert/tests/base.py"
+    "--ignore=nbconvert/tests/test_nbconvertapp.py"
+  ];
+
+
```
[the remark this answers is not in the document]
Comment (jonringer) [reply pkgs/development/python-modules/nbconvert/default.nix:57]: no

**Comment by FRidh** [comment]
the use of flit_core is making gimp fail because python2Packages.flit_core is disabled, which failing the eval as gimp is a darwin constituent https://gist.github.com/GrahamcOfBorg/f0aa782e418d7d426a7bf7a5a7372186

In principle the Python 3 flit can make a wheel for Python 2 (they're universal wheels). I've been avoiding this but can use this fact.

**Comment by FRidh** [comment]
> In principle the Python 3 flit can make a wheel for Python 2 (they're universal wheels). I've been avoiding this but can use this fact.

I had a quick look at this approach. Fun stuff! So, in principle we could run the `flit` command, Python 3 version, to build the wheel. However, e.g. `packaging` uses PEP 517, so `pip` will attempt to use that method to find the backend. Then, it fails, because it cannot find the Python 3 version. Sigh.

Really the best approach here, if we were caring sufficiently for Python 2, would be to separate the wheel building part into a separate derivation. The installation function could then simply pick the universal wheel built with Python 3. Now I would still like to do that some day, and not for Python 2, but for other reasons, but not anytime soon.

So for now I suggest we keep the older packaging for Python 2 around.

**Comment by jonringer** [comment]
what do you want to do about:
> DEPRECATION: The -b/--build/--build-dir/--build-directory option is deprecated. pip 20.3 will remove support for this functionality. A possible replacement is use the TMPDIR/TEMP/TMP environment variable, possibly combined with --no-clean. You can find discussion regarding this at https://github.com/pypa/pip/issues/8333.

TMPDIR is already set during build, so we could just remove this option if I'm realing this correctly

**Comment by groodt** [comment]
> TMPDIR is already set during build, so we could just remove this option if I'm realing this correctly

That's the same interpretation I've taken here: https://github.com/NixOS/nixpkgs/pull/105540

**Comment by FRidh** [comment]
Thank you @jonringer. There are about 900 failures left, which is about the same as it was before. I've updated to the latest pip.

**Comment by TredwellGit** [comment]
After this was merged on staging:
```
$ nixos-rebuild build -I nixpkgs=.
building Nix...
building the system configuration...
error: pytest-xdist-2.1.0 not supported for interpreter python2.7
(use '--show-trace' to show detailed location information)
```

**Comment by TredwellGit** [comment]
Seems to be caused by TeX Live depending on Python 2.

**Comment by TredwellGit** [comment]
I have TeX Live working with Python 3; will request pull later.

**Pull request merged**